File: api/src/service/portfolio.py
import random
from src.data.schema.portfolio import PortfolioResponse, Portfolio as PortfolioSchema
from src.util.log import get_logger
from fastapi import HTTPException

logger = get_logger()

class PortfolioService:
    def __init__(self) -> None:
        pass

    @staticmethod
    def holdings(data) -> PortfolioResponse:
        try:
            portfolio_mock_1 = {
                "tradingsymbol": "GOLDBEES",
                "exchange": "BSE",
                "isin": "INF204KB17I5",
                "quantity": 2,
                "authorised_date": "2021-06-08 00:00:00",
                "average_price": 40.67,
                "last_price": 42.47,
                "close_price": 42.28,
                "pnl": 3.5999999999999943,
                "day_change": 0.18999999999999773,
                "day_change_percentage": 0.44938505203405327
            }
            portfolio_mock_2 = {
                "tradingsymbol": "IDEA",
                "exchange": "NSE",
                "isin": "INE669E01016",
                "quantity": 5,
                "authorised_date": "2021-06-08 00:00:00",
                "average_price": 8.466,
                "last_price": 10,
                "close_price": 10.1,
                "pnl": 7.6700000000000035,
                "day_change": -0.09999999999999964,
                "day_change_percentage": -0.9900990099009866
            }
            return PortfolioResponse(
                status="success",
                data=[
                    PortfolioSchema(
                        **portfolio_mock_1
                    ),
                    PortfolioSchema(
                        **portfolio_mock_2
                    )
                ]
            )
        except Exception as err:
            logger.error(err, exc_info=True)
            raise HTTPException(status_code=500, detail="Error while fetching portfolio details")

File: ui/src/Portfolio.js
import * as React from 'react';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import Typography from '@mui/material/Typography';

export default function Portfolio() {

    const [rows, setRows] = React.useState([]);
    
    React.useEffect(() => {
        fetch(`api/portfolio/holdings`, {
            method: 'get',
            headers: new Headers({
              'X-Token': window.localStorage.getItem("trade-token")
            })
          })
          .then( res => res.json())
          .then(
            (result) => {
              if ("status" in result && result["status"] === "success"){
                setRows(result["data"])

              } else {
                alert("Error while fetching profile")
              }
            }, 
            (error) => {
              console.log(error);
            }
          )
}, [rows])

    
  return (
    <TableContainer component={Paper} sx={{borderStyle:"solid", margin:2, padding:2}}>
        <Typography variant='h4'>Profile Information</Typography>
      <Table sx={{ minWidth: 650 }} aria-label="simple table">
        <TableHead>
          <TableRow>
            <TableCell>Trading Symbol</TableCell>
            <TableCell align="right">Exchange</TableCell>
            <TableCell align="right">ISIN</TableCell>
            <TableCell align="right">Quantity</TableCell>
            <TableCell align="right">Authorised Date</TableCell>
            <TableCell align="right">Average Price</TableCell>
            <TableCell align="right">Last Price</TableCell>
            <TableCell align="right">Close Price</TableCell>
            <TableCell align="right">Profit and Loss</TableCell>
            <TableCell align="right">Day Change</TableCell>
            <TableCell align="right">Day Change (%)</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {rows.map((row) => (
            <TableRow
              key={row.tradingsymbol}
              sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
            >
              <TableCell component="th" scope="row">
                {row.tradingsymbol}
              </TableCell>
              <TableCell align="right">{row.exchange}</TableCell>
              <TableCell align="right">{row.isin}</TableCell>
              <TableCell align="right">{row.quantity}</TableCell>
              <TableCell align="right">{row.authorised_date.split('T')[0]}</TableCell>
              <TableCell align="right">{row.average_price.toFixed(2)}</TableCell>
              <TableCell align="right">{row.last_price.toFixed(2)}</TableCell>
              <TableCell align="right">{row.close_price.toFixed(2)}</TableCell>
              <TableCell align="right">{row.pnl.toFixed(2)}</TableCell>
              <TableCell align="right">{row.day_change.toFixed(2)}</TableCell>
              <TableCell align="right">{row.day_change_percentage.toFixed(2)} %</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
}
